### DISCUSSION TITLE
chapter_heap/build_heap/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#387] [Announcements]
# chapter_heap/build_heap/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_heap/build_heap/

**Comment by andrewperry01** [comment]
一个个插入，并保证堆结构，自上而下创建堆，是威廉姆斯提出的建堆算法时间复杂度为O(n*logn)；从最后一个非叶子结点开始构建堆，不断进行堆化，自下而上创建堆，是弗洛伊德提出来的建堆算法，时间复杂度为O(n)

**Comment by illusionno** [comment]
1.借助入堆操作实现建堆："从底至顶”堆化，堆是“自上而下”构建的，建堆时间复杂度为O(nlogn)
2. 遍历堆化实现建堆: （类似堆顶元素的出堆的后续操作）是"从顶至底"堆化，堆是“自下而上”构建的，建堆时间复杂度为O(n)

**Comment by tstanon** [comment]
大顶堆的初始化、建立、插入、出堆、释放的C语言版本代码
```c

void initHeap(float **arr, int *size) { *arr = (float *)malloc(sizeof(float) * MAX_SIZE); *size = 0; } // 初始化堆的大小为0
void peekTop(float *arr, int size, float *x) { if (size <= 0) return ; *x = arr[0]; }
void getParent(int i, int *p_idx) { *p_idx = (i - 1) / 2; }
void getLeftChild(int i, int *lc_idx) { *lc_idx = 2 * i + 1; }
void getRightChild(int i, int *rc_idx) { *rc_idx = 2 * i + 2; }
void heaptifyUp(float *arr, int idx){
    int cur = idx;
    while (cur > 0){  // 只要 cur 不是根节点
        int par; getParent(cur, &par); if (arr[cur] < arr[par]) break; // 如果当前节点小于父节点，退出循环
        float tmp = arr[cur]; arr[cur] = arr[par]; arr[par] = tmp; cur = par;  // 交换当前节点和父节点
    }
}
void heaptifyDown(float *arr, int size, int idx){
    int cur = idx;
    while (true){
        int lc, rc, maxidx = cur; getLeftChild(cur, &lc); getRightChild(cur, &rc);
        if (lc < size && arr[lc] > arr[maxidx]) { maxidx = lc;} // 找到当前节点和左子节点中的最大值
        if (rc < size && arr[rc] > arr[maxidx]) { maxidx = rc;} // 找到当前节点和右子节点中的最大值
        if (maxidx == cur) break; // 如果当前节点是最大值，退出循环
        float tmp = arr[cur]; arr[cur] = arr[maxidx]; arr[maxidx] = tmp; cur = maxidx; // 交换当前节点和最大值节点
    }
}
void pushHeap(float *arr, int *size, float x){
    if ((*size) >= MAX_SIZE) return; // 如果堆已满，直接返回
    arr[*size] = x; (*size)++; // 将新元素放在堆的最后一个位置
    heaptifyUp(arr, (*size) - 1); // 重新调整堆
}
void popHeap(float *arr, int *size, float *val){
    if ((*size) <= 0) return; // 如果堆为空，直接返回
    *val = arr[0];
    arr[0] = arr[(*size) - 1]; // 将堆的最后一个元素放在堆顶
    (*size)--;
    heaptifyDown(arr, *size,0); // 重新调整堆
}
void freeHeap(float **arr) { if (*arr) { free(*arr); *arr = NULL; } }
```

**Reply by tstanon** [reply]
借助入堆操作实现

**Comment by tstanon** [comment]
通过堆化遍历来建立堆的C代码：
```c
void buildHeap(float *arr, int *heap_size,  float *x, int size){
    memcpy(arr, x, sizeof(float) * size); *heap_size = size; // 将数组x的元素复制到堆中
    for (int i = size / 2 - 1; i >= 0; i--) heaptifyDown(arr, size, i); // 从最后一个非叶子节点开始，向上调整堆
}
```

**Reply by tstanon** [reply]
buildHeap使用heaptifyDown的原因是，如果使用 heaptifyUp从树的底部向上调整，每个节点在最坏情况下可能需要一直移到树的根部(并且底部的节点数量多)。这意味着可能需要执行更多的比较和交换操作。而如果我们从上往下调整，每个节点最多只需要向下移动几层（通常是树的高度），这使得整体效率非常高。

**Reply by tstanon** [reply]
更新，不使用额外空间。
```c
void buildHeap(float *arr, int size){
    // 使用heaptifyDown的原因是，如果使用 heaptifyUp从树的底部向上调整，每个节点在最坏情况下可能需要一直移到树的根部(并且底部的节点数量多)。
    // 这意味着可能需要执行更多的比较和交换操作。而如果我们从上往下调整，每个节点最多只需要向下移动几层（通常是树的高度），这使得整体效率非常高。
    for (int i = size / 2 - 1; i >= 0; i--) heaptifyDown(arr, size, i); // 从最后一个非叶子节点（size/2 - 1）开始，向上调整堆;
}
```

**Comment by 0101qwe** [comment]
为什么自下而上就能保住节点下面的一定是堆，如果列表前面有更小的，那不还是要移动到下面去，自上而下，列表前面如果有大的他又要重新排过，我就想知道这两种，为什么自下而上更好一点，还是有一点不理解

**Reply by HarrryHe** [reply]
从底部开始堆化可以保证每次堆化后的子堆已经是有效堆，并且从下而上能避免重复堆化，提高效率

**Reply by batreeon** [reply]
单个叶子节点一定是有效的堆，这样借助底层现成的堆一步步往上建不就很方便吗

**Reply by jiuerya7** [reply]
因为如果是按照第一种方式进行建堆操作（自顶向下），那么堆顶元素的堆化次数一定是最少的，而且每一层的每个元素的堆化次数，都比上一层的每个元素的堆化次数要多，这样下去，由于堆底层的元素数量最多，而且每一个节点执行的堆化次数都是最多的。

相反，如果采用第二种方式，先全部入堆，再从底部的  第一个非叶节点开始   进行堆化操作。
由于底层的节点数量最多，但是每个节点的堆化操作次数是最少的，只有堆顶这一个节点执行的堆化操作是最多的，因此很大程度上降低了堆化次数，也就降低了时间复杂度。

注意，前者的每一个元素都是自底向上的堆化操作，而后者的每一个元素都是自顶向下的堆化操作

**Comment by leonlovecode** [comment]
> 假设完全二叉树的节点数量为 n ，则叶节点数量为 (n+1)/2 ，其中 / 为向下整除。因此需要堆化的节点数量为 (n−1)/2 。

@krahets

请问上述说法是否有误？需要说明仅在节点数量为奇数时才适用吧（只是一个计算问题，不影响最终的复杂度）

例如一个小顶堆数组：[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

图片参考之前树的章节7.3 - [图 7-15   完全二叉树的数组表示](https://www.hello-algo.com/chapter_tree/array_representation_of_tree.assets/array_representation_complete_binary_tree.png)

图中很容易看出来，全部节点数 n = 10，其中叶节点 5 个，需要堆化的“内部节点（有子节点）”也是 5 个。

通过询问 chatgpt 数学证明，对于数组表示的 0 基（根结点下标为 0）完全二叉树而言，节点 i 的：
 - 内部节点(有孩子)的下标都满足 2i+1 < n；  
 - 内部节点数 = ⌊n/2⌋（向下取整）；  
 - 叶节点数 = ⌈n/2⌉（向上取整）；  
 - 最后一个有孩子的节点下标 = ⌊(n−2)/2⌋（向下取整）；  
 - 需要堆化的节点数，也就是全部内部节点数 = ⌊n/2⌋（向下取整）。  

上述结论也可以通过前面的例子验证，n = 10，需要堆化的节点数为 ⌊n/2⌋ = ⌊10/2⌋ = 5，正确。

若按照本章节文中的结论 `(n−1)/2` 则结果为 (10-1)/2 = 4，错误。

**Reply by sunshinesDL** [reply]
“内部节点(有孩子)的下标都满足 2i+1 < n”，直到这句还能理解，
按这个道理，非叶子结点的数量不应该等于（n-1）// 2 + 1 吗，为啥是 n // 2呢

**Reply by leonlovecode** [reply]
> 内部节点(有孩子)的下标都满足 2i+1 < n

上面这句，等价于：内部节点(有孩子)的下标都满足 i < (n-1) / 2，下面分奇偶讨论：

- 对于奇数的 n 有：**最大**的内部节点（非叶子节点）i = (n-1) / 2 - 1 或 (n-3) / 2
- 对于偶数的 n 有：**最大**的内部节点（非叶子节点）i = (n-2) / 2

统一奇偶情况，**最大**的内部节点（非叶子节点）i = (n-2) // 2 或 n // 2 - 1。因为下标从 0 开始，则 0 到 i 总共有 i+1 个非叶子节点，也即非叶子节点的数量为 (n-2) // 2 再加一，得到 n // 2

**Comment by Zhaopudark** [comment]
个人理解：

- $T(h)=2^{h+1}-h-2$
- 且 $n=2^{h+1}-1$ 即 $h+1 = \log_2(n+1)$
- 所以 $T(n)=2^{\log_2(n+1)}-1-\log_2(n+1)=n-\log_2(n+1)=O(n)$

这样理解 只需要最后部分取使用前面给出的渐近上界的定义，不必担心教程中没明确给出的基于 $O(\cdot)$ 的变量代换。

**Reply by jiuerya7** [reply]
感谢

**Comment by xiaozhuge123** [comment]
day07

**Comment by yixiu2332** [comment]
第二种建堆方法不是说自下而上堆化吗，为什么用siftDown，而不是siftUp。

**Reply by zhuiguang49** [reply]
遍历节点的时候是自下向上（也即所谓的层序遍历的倒序），但是对每个节点作堆化的时候，其实是对以该节点为根节点的子树进行从上向下（自顶向底）的堆化

**Reply by Randomlyu** [reply]
很简单，自己随便编一个完全二叉树，你用shiftUp试试不就知道哪里有问题了，你会发现用shiftUp构建不出来大（小）顶堆。

**Reply by RayneYael** [reply]
自底开始siftUp的话，想象一下，开始前，当前叶子节点到根节点的那条路径并没有满足大堆顶的要求，你对叶子节点用siftUp相当于从叶子节点开始将沿途遇到的一个max节点推到堆顶，就算你把叶子节点推上去了，能保证原叶子节点位置的数是最小值吗？并不能。

你可以用这个初始化堆手算一下: [1, 6, 3, 2, 7, 4]

用自底向上的siftUp，你第一步就会发现问题，即从最后一个叶子节点（4）开始执行，确实能把4推到堆顶，但4原本的位置是被3取代了，3原本的位置被1取代了，导致1是3的父节点，不符合大堆顶条件。

进一步，你会发现，**不管是从顶开始推，还是从底开始推，都务必朝着已经排序好（即已满足大or小子堆的）的方向去推**，否则推了也是无效的。所以自顶向下要向上推；自底向上要向下推。

**Comment by DHuayue** [comment]
For building a heap:

from the global tree perspective, the process is bottom-up, from the last internal node to the root.

From the single-node perspective, each heapify operation uses sift-down —
that is, it repeatedly compares a parent node with its children and moves the parent downward until the heap property is restored in that subtree.

This method is efficient because:

We only need to call heapify on internal nodes, not on the leaves.

Leaves are already valid heaps (no children to compare).

Therefore, we save almost half of the nodes from being traversed or checked or adjusted.
